feat: implement Serde (de)serializer for `Ipld` enum

vmx opened the pull request:
It's now possible to serialize and deserialize the `Ipld` enum with Serde.
As the Serde data model doesn't support custom types, we use the newtype
struct hack, which is used by many other libraries as well. We introduce
a newtype struct with a mostly unique/hard to clash with name. This struct
wraps the CID.

Serde de(serializers) that support CIDs, can then make use of it, others
will see a strange looking struct in their ouput.

vmx committed:
feat: implement Serde (de)serializer for `Ipld` enum

It's now possible to serialize and deserialize the `Ipld` enum with Serde.
As the Serde data model doesn't support custom types, we use the newtype
struct hack, which is used by many other libraries as well. We introduce
a newtype struct with a mostly unique/hard to clash with name. This struct
wraps the CID.

Serde de(serializers) that support CIDs, can then make use of it, others
will see a strange looking struct in their ouput.

Stebalien commented on core/Cargo.toml:
```diff
@@ -18,7 +19,10 @@ core2 = { version = "0.4", default-features = false, features = ["alloc"] }
 multihash = { version = "0.16.0", default-features = false, features = ["alloc"] }
 
 multibase = { version = "0.9.1", default-features = false, optional = true }
+serde = { version = "1.0.132", default-features= false, features = ["alloc"], optional = true }
```
ubernit:

```suggestion
serde = { version = "1.0.132", default-features = false, features = ["alloc"], optional = true }
```

Stebalien commented on core/src/error.rs:
```diff
@@ -36,6 +36,26 @@ pub struct InvalidMultihash(pub Vec<u8>);
+
+#[cfg(feature = "serde-codec")]
+impl serde::de::Error for SerdeError {
+    fn custom<T: std::fmt::Display>(msg: T) -> Self {
```
`core::fmt`

Stebalien replied:
It looks like we may need nostd tests.

vmx replied:
Currently building `cargo build --no-default-features --features serde-codec --target thumbv6m-none-eabi --manifest-path core/Cargo.toml` fails due to Serde. So that might be difficult to test.

Stebalien replied:
Hm. Just building with `#![no_std]` should catch this, even without `thumbv6m-none-eabi`.

Stebalien commented on core/src/ipld.rs:
```diff
@@ -35,6 +35,21 @@ pub enum Ipld {
+// values, hence it's OK to implement `Eq`.
+impl Eq for Ipld {}
+
+// TODO vmx 2021-12-21: Make sure that special float values like Nan or Infinity are never
```
This is also an issue for Eq. Will we have cases that can panic?

Stebalien commented on core/src/ipld.rs:
```diff
@@ -7,13 +7,13 @@ use alloc::{
 
 /// Ipld
-#[derive(Clone, PartialEq)]
+#[derive(Clone, PartialEq, PartialOrd)]
```
Have we defined ordering somewhere?

- Integer/Float? It should be possible to order between these, but do we want to?
- What about maps? I assume maps simply can't be compared?
- Do we define a "null is less than everything" ordering?

I'd consider punting on this unless we really need it.

Stebalien replied:
I guess this derive will order based on the order in the enum. That's probably not what we want.

Stebalien commented on core/src/ipld.rs:
```diff
@@ -35,6 +35,21 @@ pub enum Ipld {
+impl Ord for Ipld {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other)
+            .expect("NaN and Infinity are not supported by IPLD.")
```
This seems like a pretty big gotcha given that users can construct this value.

Stebalien commented on core/src/lib.rs:
```diff
@@ -1,6 +1,6 @@
 //! Core ipld types used by ipld codecs.
 #![deny(missing_docs)]
-#![deny(warnings)]
+//#![deny(warnings)]
```
delete?

Stebalien commented on core/src/serde/mod.rs:
```diff
@@ -0,0 +1,140 @@
+/// Utility for testing (de)serialization of [`Ipld`].
+///
+/// Checks if `data` and `ipld` match if they are encoded into each other.
+pub fn assert_roundtrip<T>(data: &T, ipld: &Ipld)
```
Should this really be public if it's for testing?

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,454 @@
+        S: ser::Serializer,
+    {
+        match &self {
+            Self::Null => serializer.serialize_unit(),
```
Shouldn't this be `serialize_none`?

vmx replied:
That's also what `serde_json` is doing:
https://github.com/serde-rs/json/blob/7025523603fe604d11b92ccd4ca314e343d3ae50/src/value/ser.rs#L22

And for me personally Rust's `None` kind of also implies there's a `Some`, which obviously isn't the case for `null`.

vmx replied:
> And for me personally Rust's `None` kind of also implies there's a `Some`, which obviously isn't the case for `null`.

Which actually is the case if you use `Ipld::Null`, so it makes sense representing it as `None`.

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,454 @@
+            Self::Null => serializer.serialize_unit(),
+            Self::Bool(bool) => serializer.serialize_bool(*bool),
+            Self::Integer(i128) => serializer.serialize_i128(*i128),
+            Self::Float(f64) => serializer.serialize_f64(*f64),
```
nit: avoid using type names as variable names, it's a bit confusing.

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,454 @@
+
+    #[inline]
+    fn serialize_i16(self, value: i16) -> Result<Self::Ok, Self::Error> {
+        self.serialize_i64(i64::from(value))
```
why not directly to i128?

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,454 @@
+
+    #[inline]
+    fn serialize_unit(self) -> Result<Self::Ok, Self::Error> {
+        Ok(Self::Ok::Null)
```
Honestly, I'd serialize this as an empty list, not as null. Unit is `()` (emtpty tuple).

vmx replied:
It's what `serde_json` is doing: https://github.com/serde-rs/json/blob/7025523603fe604d11b92ccd4ca314e343d3ae50/src/value/ser.rs#L160-L162 To me it makes sense that we do the same.

Stebalien replied:
See below.

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,454 @@
+
+    #[inline]
+    fn serialize_unit_struct(self, _name: &'static str) -> Result<Self::Ok, Self::Error> {
+        self.serialize_unit()
```
ditto above. That or, really we could just _reject_ unit.

Stebalien replied:
Alternatively, it would make sense to serialize this as the name itself (as a constant). But... I'd still prefer to just not implement it.

vmx replied:
Same again, that's what `serde_json` is doing: https://github.com/serde-rs/json/blob/7025523603fe604d11b92ccd4ca314e343d3ae50/src/value/ser.rs#L165-L167

Stebalien replied:
Serde is too automagical, IMO. I can't think of why I, as a user, would want unit and/or a unit struct serialized to "null" automatically.

Stebalien replied:
Basically, serde's goal is to "just work". But with IPLD, we really care about the specifics.

Stebalien replied:
Actually, let's just have this discussion in the comments.

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,454 @@
+        _variant_index: u32,
+        variant: &'static str,
+    ) -> Result<Self::Ok, Self::Error> {
+        self.serialize_str(variant)
```
Honestly, I'd expect this to use the variant index, at least in rust.

Maybe we should just not implement this? I.e., force the user to specify their representation when they derive serialize on their type?

vmx replied:
This is again what `serde_json` is doing: https://github.com/serde-rs/json/blob/7025523603fe604d11b92ccd4ca314e343d3ae50/src/value/ser.rs#L165-L167

Stebalien replied:
See above. Especially in this case where, personally, I'd expect it to encode the enum as the underlying representation. E.g.:

```rust
#[repr(u8)]
enum Foo {
    Yes = 1,
    No = 2,
}
```

Stebalien replied:
Eh, discuss below.

Stebalien reviewed (commented):
Initial review. My main feedback would be to reject more types, forcing the user to specify the representation.

vmx committed:
chore: add space

Co-authored-by: Steven Allen <steven@stebalien.com>

vmx committed:
fix: use core instead of std

vmx committed:
fix: remove Ord, PartialOrd and Eq from Ipld enum

The float type is the problem here. Although the IPLD data model states,
that only numbers are allowed, it's hard to prehibit non number values.
Hence don't implement those for now.

vmx committed:
fix: no need to make assert_roundtrip public

vmx committed:
chore: accidentally commented out

Remove that in a different PR.

vmx committed:
refactor: don't use type names for variable names

vmx committed:
refactor: simplify serializer

vmx committed:
fix: make `serde-codec` feature compile with no_std

vmx committed:
refactor: use nicer BTreeMap syntax

vmx committed:
fix: serialize `Ipld::Null` as `None`

This matches closer the intention of `Ipld:Null`, which is usually used of
some value isn't set. Hence it's more like a `Option`, rather than a unit
type.

vmx committed:
fix: error on unit types

This commit also adds some documentation on how Rust types are serialized.

vmx committed:
fix: don't allow deserializtion of unit (struct)

Make the deserialization symmetrical to the serialization.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+            where
+                V: de::SeqAccess<'de>,
+            {
+                let mut vec = Vec::new();
```
nit: use `visitor.size_hint()`.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+
+            /// Newtype structs are only used to deserialize CIDs.
+            #[inline]
+            fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
```
We should test this with non dag codecs to make sure we don't get any strange results (e.g, infinite recursion).

Stebalien replied:
See below, I think this is backwards. If we're told to visit a newtype struct, I think we need to turn around and try to deserialize into a CID. `BytesToCidVisitor` will deserialize any bytes into a CID.

vmx replied:
I still feel like I'm missing something. Are we talking about a call like this:

```rust
Ipld::deserialize(something_that_impls_Deserializer);
```

You'd want to deserialize from some codec into `Ipld`. This would usually be some blob that implements a DAG codec. So now your question is: "could we get any strange results, when some random `Deserializer` is used?". I don't know why you would do that, but let's say you do.

Your `Deserializer` implementation would need to have a code path that calls `visit_newtype_struct`. Then that `Deserializer` also needs to have a `deserialize_bytes` implementation. That could of course call any `visit_*` it likes. But luckily we pass in the [`BytesToCidVisitor`](https://github.com/multiformats/rust-cid/blob/6cf0f6768e5b9e12f802e23bb63673d729898eef/src/serde.rs#L38-L68), which only implements `visit_bytes` and `visit_seq`. So all other `visit_*` calls would error, no harm here. When it calls `visit_bytes` or `visit_seq` we get some bytes or a sequence. If those form by coincidence a valid CID, we would return that CID. Else we would again error.

To conclude: a random `Deserializer` implementation is likely to error. Only if it leads to a byte sequence that leads to a valid CID, a CID is returned. And I would say if it does, it was likely a codec that was crafted to support CIDs.

Stebalien replied:
So "random bytes" aren't the issue, just assume that the attacker can put bytes forming a CID anywhere. The constraint is that some `links` function that extracts all CIDs in some object should list the exact same CIDs we get when we decode that object.

Basically:

- Assume the programmer is well-meaning but may not have the same mental model as you.
- Assume that input data is malicious.

Stebalien replied:
I'm setting the bar to "we can prove to ourselves that this is correct" not "we can't think of a way for this to go wrong".

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+        fn $deserialize<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+            match self {
+                Self::Integer(integer) => {
+                    if integer >= $ty::MIN.into() && integer <= $ty::MAX.into() {
```
consider using `let int: $ty = integer.try_into()?`?

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+
+    #[inline]
+    fn deserialize_any<V: de::Visitor<'de>>(self, _visitor: V) -> Result<V::Value, Self::Error> {
+        unreachable!()
```
What if we use `Ipld` with, e.g., non-dag json? I think we need to return a type error instead of panicing.

Stebalien replied:
This also panics when attempting to decode a CID into an untagged union.

Stebalien replied:
https://github.com/ipld/libipld/pull/141

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+            Self::Bytes(bytes) => visitor.visit_bytes(&bytes),
+            Self::List(list) => visit_seq(list, visitor),
+            Self::Map(map) => visit_map(map, visitor),
+            Self::Link(cid) => visitor.visit_newtype_struct(CidDeserializer(cid)),
```
Maybe we should just fail here? I.e., won't this deserialize into any:

```rust
struct NewType(Bytes);
```

?

Stebalien replied:
Hm. Well, we need to support deserializing IPLD -> IPLD...

Stebalien replied:
Yeah, I think we need a different deserializer. One that expects a call to (and only a call to) `deserialize_newtype_struct`.

vmx replied:
I've added a few more tests (the the most recent commit). `struct Wrapper(ByteBuf)`, does *not* deserialize a CID. If you still have concerns, please provide concrete tests (can be pseudo code), what I should test for.

vmx replied:
> Hm. Well, we need to support deserializing IPLD -> IPLD...

~While I was looking into this, I was actually wondering about that. Do we really need that? Does a `deserialize_any` implementation make sense, or should we just remove it? Why would I want to deserialize `Ipld` into `Ipld`? I mean, I have an `Ipld` object already.~
`deserialized_any` is e.g. used when you'd want to deserialize `Ipld` into a `serde_json::Value`.

Stebalien replied:
Generic code? Without that, we could end up with some strange bugs.

Stebalien replied:
> I've added a few more tests (the the most recent commit). struct Wrapper(ByteBuf), does not deserialize a CID. If you still have concerns, please provide concrete tests (can be pseudo code), what I should test for.

Hm. Something still seems strange. I'll try to create a test.

Stebalien replied:
Ah, it's because we're not calling "deserialize any".

Stebalien replied:
Ok, so, to trigger this, some code would need to:

1. Call `deserialize_any`.
2. Get called by `visit_newtype_struct`.
3. Call `deserialize_bytes`.

At the moment, I can only get step 3 to call back into `deserialize_any` (which panics). I'm not sure if anyone is turning around and calling `deserialize_bytes` in this case, but there's no reason they _can't_.

However, it would still be safer to expect a call to `deserialize_newtype_struct` so we can _check_ that we're deserializing into the correct thing. I.e.:

1. This function passes a deserializer that _only_ succeeds if called with `deserialize_newtype_struct` with the special CID string.
2. `visit_newtype_struct` on the `Ipld` type calls back into `deserialize_newtype_struct`.

We won't get infinite recursion because we'd be passing a custom deserializer.

vmx replied:
> 2. `visit_newtype_struct` on the `Ipld` type calls back into `deserialize_newtype_struct`.

This is not clear to me. To which `deserialize_newtype_struct` is it calling back, to the original one?

> We won't get infinite recursion because we'd be passing a custom deserializer.

I lost count, how many deserializers would we have? Changing the `CidDeserializer` to one that accepts `deserialize_newtype_struct` only, or is this yet another one.

When I was developing this, I always hit the problem that eventually I need to call `deserialize_bytes` as I haven't found any other way to pass on the bytes of the CID into a visitor, which then needs to construct the `Cid`. Though it can well be that there is a way.

Stebalien replied:
> To which deserialize_newtype_struct is it calling back, to the original one?

A new one, unfortunately.

---

But I don't think that works. `visit_newtype_struct` really shouldn't call `deserialize_newtype_struct`. It should always be the other way around. The problem we're running into here is that we're calling `visit_newtype_struct` from `deserialize_any` where we don't have any type information.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+
+    fn deserialize_f32<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        match self {
+            Self::Float(float) => visitor.visit_f32(float as f32),
```
Hm. Don't we want some form of "fits" test?

Stebalien replied:
Maybe check validity (non-nan, etc.) here as well?

Stebalien replied:
`(float as f32) as f64 == float` should be a sufficient "fits" test.

vmx replied:
I've put it into the documentation, that floats are converted with loss. I thought one might want to serialize into a `f32`, e.g. within a struct and don't really care about the loss of precision. But I can also see the point of only converting if it fits. I don't really have a strong opinion on that one, so I just made the change.

Stebalien replied:
Personally, I'd rather force the user to specify that they're fine with a loss. They can always decorate the f32 with some custom deserializer that decodes into an f64 and casts.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+    fn deserialize_char<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        match self {
+            Self::String(string) => {
+                if string.len() == 1 {
```
Length is bytes (https://doc.rust-lang.org/std/string/struct.String.html#method.len). Don't you love utf-8?

We're better off just calling `string.chars()` and succeeding iff there's exactly one char.

vmx replied:
Good catch! It's embarrassing that I didn't catch that one myself ;)

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        match self {
+            Self::Map(map) => visit_map(map, visitor),
```
Do we need to check the fields here?

vmx replied:
Resolved, but later comment. We don't need special checks here, missing or wrong fields will error, tests are there to cover that case.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+        _len: usize,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        self.deserialize_seq(visitor)
```
Do we need to check the length?

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        drop(self);
+        visitor.visit_none()
```
You know, I wonder if this should be a `visit_unit`?

vmx replied:
I've to admit, I don't fully understand how `deserialized_ignored_any()` is actually triggered. I just don't understand its documentation. So I'm fine with `visit_unit()` if you think it's correct.

Stebalien replied:
Oh, I see. It literally doesn't matter. `deserialize_ignored_any` is just a hint to the deserializer that it shouldn't _actually_ bother decoding anything. The visitor will "go through the motions", but will ignore any results. E.g., if you ask it to visit a map, it'll do that then throw away the results.

So calling `visit_something_free` is the right way here (none, unit, etc.). Let's leave it as-is.

vmx replied:
I've changed it to `visit_unit()`, but I can also revert it.

Stebalien replied:
I'd leave it as is.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+    }
+}
+
+fn visit_map<'de, V>(map: BTreeMap<String, Ipld>, visitor: V) -> Result<V::Value, SerdeError>
+where
+    V: de::Visitor<'de>,
+{
+    let mut deserializer = MapDeserializer::new(map);
+    visitor.visit_map(&mut deserializer)
+}
+
+fn visit_seq<'de, V>(list: Vec<Ipld>, visitor: V) -> Result<V::Value, SerdeError>
```
We're dropping the size hints here.

vmx replied:
Do you mean, that there should be a `size_hint` implementation in the `de::SeqAccess` impl? This is what `serde_json` does.

Stebalien replied:
Oh, never mind. I noticed we were dropping `_len` in https://github.com/ipld/libipld/blob/ad768fe128a76c2c1ea0fcb40bbcc1ff23644ae8/core/src/serde/de.rs#L696, but we don't actually need that (the list tells us what we need).

Stebalien reviewed (changes requested):
Starting with a deserialize review. I'll review serialize next.

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,486 @@
+    where
+        T: ser::Serialize,
+    {
+        let values = BTreeMap::from([(variant.to_owned(), value.serialize(self)?)]);
```
uber-nit: insert will be faster (this will get collected into a temporary array otherwise).

vmx replied:
I hoped that it would be the reverse and that the compiler could optimize that one even better.

Stebalien replied:
Allocations (side effects) can make optimizations a bit difficult, but I haven't tested.

Stebalien commented on core/src/serde/ser.rs:
```diff
@@ -0,0 +1,486 @@
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        let map = BTreeMap::from([(self.name, Self::Ok::List(self.vec))]);
```
(insert would be faster)

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,693 @@
+            {
+                let mut values = BTreeMap::new();
+
+                while let Some((key, value)) = visitor.next_entry()? {
+                    values.insert(key, value);
+                }
```
uber-nit: in cases like this, it might be faster to collect into a vector first, then construct the map from `vec.into_iter()`. Internally, the BTreeMap will:

1. Collect that iterator back into a `vec`, which is a no-op due to specialization.
2. Sort it.
3. Perform "bulk" insertion to construct the tree all at once.

Of course, that might not really be worth it. Up to you.

Stebalien reviewed (commented):
nits

vmx committed:
fix: address code review

Mostly better errors and more tests.

vmx reviewed (commented):
I've skipped the serializer nits for now. I guess I'll just leave them like that for now (I'm in the "optimize when things work and bottlenecks are identified camp :). Though it's always great to hear about such possible improvements.

One big thing I haven't addressed with the most recent changes is the `CidDeserializer` and the comments around whether things would work with a non-dag Serde codec. I'll look into those tomorrow.

vmx committed:
fix: update no longer accurate comment

vmx committed:
test: add additional tests for CIDs

Make sure that CIDs are never deserialized to bytes.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,741 @@
+        V: de::Visitor<'de>,
+    {
+        match self.0 {
+            Some(Ipld::List(v)) => visit_seq(v, visitor),
```
Check that `_len == v.len()`?

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,741 @@
+        _len: usize,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        self.deserialize_seq(visitor)
```
call `self.deserialize_tuple` and pass the length.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,741 @@
+            Self::Map(map) => {
+                let keys: Vec<_> = map.keys().collect();
+                let mut fields_sorted = fields.to_vec();
+                fields_sorted.sort_unstable();
```
So... I think I might have lied here. I _think_ we need to just treat the fields as a hint. Otherwise, we can't accept objects with extra fields. The decision to allow unrecognized fields should lie with the visitor.

vmx replied:
Cool, then this can be removed as Serde takes care of erroring in the other case like not matching or missing fields.

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,741 @@
+                    }
+                } else {
+                    error(format!("`Ipld::Float` must be a finite number, not infinity or NaN, input was `{}`", float))
+                }
```
nit:

```suggestion
                if !float.is_finite() {
                    error(format!("`Ipld::Float` must be a finite number, not infinity or NaN, input was `{}`", float))
                } else if (float as f32) as f64 != float {
                    error("`Ipld::Float` cannot be deserialized to `f32`, without loss of precision`")
                } else {
                    visitor.visit_f32(float as f32)
                }
```

Stebalien commented on core/src/serde/de.rs:
```diff
@@ -0,0 +1,741 @@
+        let (variant, value) = match self {
+            Ipld::Map(map) => {
+                let mut iter = map.into_iter();
+                let (variant, value) = match iter.next() {
+                    Some(v) => v,
+                    None => {
+                        return error(
+                            "Only `Ipld::Map`s with a single key can be deserialized to `enum`, input had no keys"
+                        );
+                    }
+                };
+                // Enums are encoded in IPLD as maps with a single key-value pair
+                if iter.next().is_some() {
+                    return error(
+                        "Only `Ipld::Map`s with a single key can be deserialized to `enum`, input had more keys"
+                    );
+                }
```
ot: what I wouldn't give to have a `try_collect`:

```rust
let [(variant, value)] = map.into_iter().try_collect()?;
```

vmx committed:
test: add tests with serde_json::Value

Make sure that deserializing CIDs lead to an error.

Stebalien committed:
fix: CidDeserializer: error instead of panic

Thanks @stebalien for the test case.

vmx committed:
refactor: better if-else

Co-authored-by: Steven Allen <steven@stebalien.com>

vmx committed:
fix: check for all tuple sizes

vmx committed:
fix: allow additional fields when deserializing structs
